### ISSUE TITLE
Blocks in the orphanage are treated as novel

### ISSUE THREAD

**Issue by geigerzaehler** [oscoin/oscoin#464]
`Oscoin.Storage.applyBlock` should return `Stale` when the block has already been seen by the node. This includes blocks that are in the orphanage. But `applyBlock` does not have access to the Orphanage. It only has access to `BlockStoreReader` and `BlockStoreReader` only contains the blocks forming the best block chain. As a result `applyBlock` will return `Applied` even if a block is in the orphanage. This in turn prevents us from using `Protocol` in the consensus network tests.

**Comment by adinapoli-mndc** [comment]
Nice catch. The `orphanage` stores the full set of block hashes it encountered, so it shouldn't be hard to extend the `Protocol` to also perform this check. The first thing which comes to mind is to extend the `Protocol` with a `isNovel` function, and _remove_ the `BlockStore.isNovel` function as that has little sense holistically speaking.

**Comment by geigerzaehler** [comment]
I think this makes sense. `Protocol.isNovel` would check the block store as well as the orphange. The question is how we provide this to `applyBlock` do we add `isNovel` to `Procol.Handle` and pass the handle to `applyBlock`?

**Comment by geigerzaehler** [comment]
> I think that's the easiest and cleanest solution. Can you think of any drawback / impediment in doing that?

I think it is not ideal and requires a bit of reshuffling but it seems to be the best solution.

**Comment by adinapoli** [comment]
> I think it is not ideal and requires a bit of reshuffling but it seems to be the best solution.

Do you want to discuss this in a call later today? On top of my head I am surprised you say so 😮 When we create the `Handle` in `runProtocol`, we have in scope the `MVar Protocol` function which is kept up-to-date every time we run `stepProtocol`, so we could in principle define something like:

```
              { dispatchBlockSync = \blk -> do
                  events <- modifyMVar proto $ \p -> stepProtocol p blk
                  forM_ events (Telemetry.emit telemetry)
              , dispatchBlockAsync = atomically . writeTBQueue incomingBlocksQueue
              , isNovel = withMVar proto $ \p -> ...
              }
```

What am I missing? 🤔 (mind, didn't have morning ☕️ yet 😁 )

**Comment by adinapoli-mndc** [comment]
> I think it is not ideal and requires a bit of reshuffling but it seems to be the best solution.

Do you want to discuss this in a call later today? On top of my head I am surprised you say so 😮 When we create the `Handle` in `runProtocol`, we have in scope the `MVar Protocol` function which is kept up-to-date every time we run `stepProtocol`, so we could in principle define something like:

```
              { dispatchBlockSync = \blk -> do
                  events <- modifyMVar proto $ \p -> stepProtocol p blk
                  forM_ events (Telemetry.emit telemetry)
              , dispatchBlockAsync = atomically . writeTBQueue incomingBlocksQueue
              , isNovel = withMVar proto $ \p -> ...
              }
```

What am I missing? 🤔 (mind, didn't have morning ☕️ yet 😁 )

**Comment by adinapoli-mndc** [comment]
(Sorry, I had to delete the previous comment as I've accidentally commented with my doppelgänger 😅

**Comment by cloudhead** [comment]
The only other option I see would be to have a `BlockTree` module which combines block storage with chain selection and orphan storage.

**Comment by adinapoli-mndc** [comment]
> The only other option I see would be to have a BlockTree module

You are really attached to this `BlockTree` idea, aren't you? 😁 😂

**Comment by cloudhead** [comment]
I mean, we are talking about a tree of blocks with some orphans, :thinking: :roll_eyes: it makes sense for the best chain to just be a pointer to a tip (leaaf) of a tree :man_shrugging:

**Comment by adinapoli-mndc** [comment]
Yes, I see what you mean! I guess my idiosyncrasy for a `BlockTree` (well, that's not technically right, I am not opposed to a `BlockTree` module at all, I am just trying to reason out loud and convince me it does fit) is that the way I see things, the `BlockTree` is a dumb collection of things (store + orphanage) but the chain selection is something which belongs to somebody which has a "higher" knowledge of the system (i.e. the `Protocol`).

If you think an acceptable design would be for the `Protocol` to still "step the world" and the `BlockTree` to be this dumb container of blocks + orphans, I think I quite like it. 

Under this scenario one could ask the `BlockTree` for "is this block novel?" rather than going via the `Protocol`. Question remains whether or not the miner should _also_ know the `BlockTree` as at the moment it's already coupled with the `Protocol.Handle` _anyway_, so passing a handle to a `BlockTree`, in this sense, would make the system more coupled.

What do you guys think?

**Comment by cloudhead** [comment]
That's more or less how I think about it, something like:
* The BlockTree just stores the tree of blocks, + blocks with no parent (orphan)
* The protocol holds a "pointer" to the best chain in the BlockTree (essentially just a tip or a hash)
* The BlockTree knows if a block is novel, since it has all known blocks
* The BlockTree can tell you which chain is its best, according to an externally-supplied scoring function (it could also store it internally for optimization purposes)

**Comment by adinapoli-mndc** [comment]
Me and @geigerzaehler discussed this over the wire and we both feel that while the `BlockTree` might conceptually make sense, but for our application the path of least resistance is to go with adding the `isNovel` function to the `Protocol.Handle`, for a couple of reasons:

* Putting `isNovel` inside a `Protocol.Handle` means that now we don't need `applyBlock` to take the `BlockStore` (nor the `BlockTree`) as input, which simplify the system and make it less coupled. If we introduce the `BlockTree` now `applyBlock` needs to depend on more things;

* A "pure" `BlockTree` is hard to write, because our `Orphanage` is completely pure but we need a data structure that keeps it up-to-date, and as of today, this data structure is the `Protocol.Handle`, which lives in `IO` and use an `MVar` to keep track of the updates to the `Orphanage`. In summary, our `BlockTree` would end up looking essentially like the `Protocol` , and the protocol would be at that point vacuous (I don't think that keeping a pointer around would justify a new module/data structure).

We do promise to keep an open mind and to consider again the `BlockTree` would the design call for it 😁
